sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -2025,12 +2077,34 @@ pi_result cuda_piextContextGetNativeHandle(pi_context context,
 /// \param[out] context Set to the PI context object created from native handle.
 ///
 /// \return TBD
-pi_result cuda_piextContextCreateWithNativeHandle(pi_native_handle, pi_uint32,
-                                                  const pi_device *, bool,
-                                                  pi_context *) {
-  cl::sycl::detail::pi::die(
-      "Creation of PI context from native handle not implemented");
-  return {};
+pi_result cuda_piextContextCreateWithNativeHandle(pi_native_handle nativeHandle,
+                                                  pi_uint32 num_devices,
+                                                  const pi_device *devices,
+                                                  bool ownNativeHandle,
+                                                  pi_context *piContext) {
+  (void)num_devices;
+  (void)devices;
+  (void)ownNativeHandle;
+  assert(piContext != nullptr);
+  assert(ownNativeHandle == false);
+
+  CUcontext newContext = reinterpret_cast<CUcontext>(nativeHandle);
+
+  ScopedContext active(newContext);
+
+  // Get context's native device
+  CUdevice cu_device;
+  pi_result retErr = PI_CHECK_ERROR(cuCtxGetDevice(&cu_device));
+
+  // Create a SYCL device from the ctx device
+  pi_device device = nullptr;
+  retErr = cuda_piextDeviceCreateWithNativeHandle(cu_device, nullptr, &device);
```

Comment:
Instead of creating a new `pi_device` here, would it make sense to simply check `cu_device` against `devices` and pass the (hopefully only) device in `devices` to the new `_pi_context`?

That is, check that `num_devices == 0` and that `cu_device == devices[0].get()`.

Reply:
I think this should stay as is. Because `devices` is always nullptr and `num_devices` is always 0. I think only the level_zero plugin uses these parameters. I have linked to the relevant code below.

Additionally `cuda_piextDeviceCreateWithNativeHandle` does not create a new `pi_device` but searches through all platforms until it finds one which has the same native_handle. This is done, so far as I understand, because the number of SYCL devices should be static throughout the applications lifetime.

`make_context` backend calls:
level_zero: https://github.com/intel/llvm/blob/sycl/sycl/source/backend/level_zero.cpp#L46-L62
otherwise: https://github.com/intel/llvm/blob/sycl/sycl/source/backend.cpp#L67-L78

Reply:
Ah, right you are.

> Additionally `cuda_piextDeviceCreateWithNativeHandle` does not create a new `pi_device` but searches through all platforms until it finds one which has the same `native_handle`. This is done, so far as I understand, because the number of SYCL devices should be static throughout the applications lifetime.

From what I remember, it is also inspired by how OpenCL handles devices, specifically that the reference count of root-devices remain unaffected by retain/release, meaning native root-device representations can stay alive behind the scenes.